apps/ledger-live-desktop/src/renderer/screens/customImage/index.tsx

```diff
@@ -0,0 +1,229 @@
+import { ImageDownloadError } from "@ledgerhq/live-common/customImage/errors";
+import { withV2StyleProvider } from "~/renderer/styles/StyleProvider";
+
+type Props = {
+  imageUri?: string;
+};
+
+const orderedSteps: Step[] = [
+  Step.chooseImage,
+  Step.adjustImage,
+  Step.chooseContrast,
+  Step.transferImage,
+];
+
+const ErrorDisplayV2 = withV2StyleProvider(ErrorDisplay);
+
+const CustomImage: React.FC<Props> = props => {
+  const { imageUri } = props;
+  const { t } = useTranslation();
+
+  const [stepError, setStepError] = useState<{ [key in Step]?: Error }>({});
+
+  const [sourceLoading, setSourceLoading] = useState<boolean>(false);
+
+  const [loadedImage, setLoadedImage] = useState<ImageBase64Data>();
+  const [croppedImage, setCroppedImage] = useState<ImageBase64Data>();
+  const [finalResult, setFinalResult] = useState<ProcessorResult>();
+
+  /**
+   * Keeping a record of the crop params of a given image so that the cropping
+   * state is not lost when unmounting the cropping step component.
+   * */
+  const [initialCropParams, setInitialCropParams] = useState<CropParams>();
+
+  const [step, setStep] = useState<Step>(Step.chooseImage);
+
+  console.log("step", step);
+
+  const initialUri = imageUri;
+
+  useEffect(() => {
+    let dead = false;
+    if (initialUri) {
+      setSourceLoading(true);
+      urlContentToDataUri(initialUri)
+        .then(res => {
+          if (dead) return;
+          setLoadedImage({ imageBase64DataUri: res });
+          setStep(Step.adjustImage);
+        })
+        .catch(e => {
+          console.error(e);
+          if (dead) return;
+          setStepError({ [Step.chooseImage]: new ImageDownloadError() });
+        });
+    }
+    return () => {
+      dead = true;
+    };
+  }, [setLoadedImage, initialUri]);
+
+  useEffect(() => {
+    if (loadedImage) setSourceLoading(false);
+  }, [loadedImage]);
+
+  const handleStepChooseImageResult: ComponentProps<
+    typeof StepChooseImage
+  >["onResult"] = useCallback(res => {
+    setLoadedImage(res);
+    setStep(Step.adjustImage);
+  }, []);
+
+  const handleStepAdjustImageResult: ComponentProps<
+    typeof StepAdjustImage
+  >["onResult"] = useCallback(res => {
+    setCroppedImage(res);
+  }, []);
+
+  const handleStepChooseContrastResult: ComponentProps<
+    typeof StepChooseContrast
+  >["onResult"] = useCallback(res => {
+    setFinalResult(res);
+  }, []);
+
+  const handleError = useCallback(
+    (step: Step, error: Error) => {
+      setStepError({ [step]: error });
+    },
+    [setStepError],
+  );
+
+  const errorHandlers: { [key in Step]: (error: Error) => void } = useMemo(
+    () => ({
+      [Step.adjustImage]: handleError.bind(null, Step.adjustImage),
+      [Step.chooseContrast]: handleError.bind(null, Step.chooseContrast),
+      [Step.chooseImage]: handleError.bind(null, Step.chooseImage),
+      [Step.transferImage]: handleError.bind(null, Step.transferImage),
+    }),
+    [handleError],
+  );
```

Comment:
It works, but IMO it was hard to read with the `bind`. Could we simplify it with `handleError` being a "higher order" function, aka:
```
const handleError = useCallback(
    (step: Step) => (error: Error) => {
      setStepError({ [step]: error });
    },
    [setStepError],
  );
```

And then on the `onError` prop:
```
onError={handleError(Step.aStep)}
```

? WDYT ? Your solution works, i just found it harder to read at first

Reply:
Yeah my solution is kinda weird indeed, I was trying to avoid redeclaring a new function in the JSX to prevent triggering useless rerenders (for instance `handleError={(error) => setStepError({[step]: error})}` . And I think your solution would do the same, it's returning a new reference at every call.
I could avoid the bind though if it looks a bit obscure.